### DISCUSSION TITLE
Base64 gcode thumbnails preview on DWIN

### DISCUSSION THREAD

**Discussion by tiagofreire-pt** [Jyers/Marlin#885] [Ideas]
I'm wondering if the DWIN screen has the capability to do somehow a preview of [base64 renders](https://www.base64decode.net/base64-image-decoder), like the Prusa printers do on their LCD.

This is a base64 example, generated by Cura 4.9.1 post processing thumbnail extension, inserted into the gcode.
[CE3_z_knob Generic PLA lw 0.38mm lh 0.16mm if 15% ext1 205C bed 60C.zip](https://github.com/Jyers/Marlin/files/6527282/CE3_z_knob.Generic.PLA.lw.0.38mm.lh.0.16mm.if.15.ext1.205C.bed.60C.zip)

![99ca18b84f1beecf1b3fea780bde91e3](https://user-images.githubusercontent.com/41837236/119242884-8e5ee800-bb59-11eb-8c32-4ed47ef14d41.png)

```iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAACXBIWXMAAA7EAAAOxAGVKw4bAAAIOU
lEQVRYhZWXa4xdVRmGn3U5e59z5j7T6RRLS6nTdkRogWJRwg+CCRESExNNjBiJiQESRY3+MSYkioLx
n79MjEg0Qg3WpBgShJpgsUIKIvRCKZSWdkLpdNqZzum5zdm3tT5/7HObUgF3snN21l5nfe/3fu9617
cVH+Pa+xNsOH39juGy3jE8YmdsUa83RTuuAlMGUKlbzuJsKWnK+9Vq9vZyM/u3jewrN93/WvpRa6sP
e3noj1tXZ9b+eLCs75kcM+OjIwYVKgg0hBZKYT4xSiBKIfFILFRrjoWKu1ht+J3Kq0du+sZrZ/9vAE
/+8tPbrl4bPitwxapRzdiQYbCsaTkh8uCVBmPyyc6hxVPUUDKKxrKnUvdcuOgAFk4vJHd95YdH/vOx
ATz83enPTq8r/W1yzIxV656pCcO5JUcUCyODmuEBTTFQWKtQQOaEVizUmp5a0xMWFFMThvNLjqGyZq
nuau+cir744K+P77s0lr4cgKmJ8InhAT02NqQxBp7e1+TQ8Zg4FURApD2x/dx/x4nwxrsJT73QRCkY
H9EMlfXw1GTw5OVi2csNhlZNjQ8bBoqa7TMhMxsC3pvPiGKhYHPSRKCLQ0ApCAqKiRHDJyYt69dYBk
ua5ciTDsHCRb/6YwN49uXl1vo1drA8aRgoaSbHNJs2BHijiD1kGDIbIDZAAO0SbJpgcYQatBOyWGhG
HkEzf8Gx5+Xl1uVifaAEe/diaw3vdr/Q5J8HIuYXHQBKg7FQHjQMT4aMbxhmYnqSVdOTjF81zPBkSH
nQYKxCaRAF55YcLx5ssXtvg4s1l+3adU3woQwc3LX9Vj/PY0PlyhrvYXYupdrwvH4sZs2kYdW4ZWjE
UByBQrOADnP8PolIKwlRLaFezbiw5Di7kFGteyp1h/NQKqrRT7rSW6/96cZ7t9/9+j8+AOD5R7d+DZ
E/HDmZBL5dXwFEBOeFStXTaqWE5zOsTTC6jlI9DTgPWSYkmRDFgnOC0CdagTdOxBu3bQ73vPDotvtu
u/fQ77sl+NkDM3eXi/bxakMCEbh+c4CIdNXer/wOqI4ALx2TFeA7t7Btc4jWcLHubLlkfvfQtzf/AM
A8/P2ZL48M6p2b1xWsc7m3XHN1yNiQptr0BFYRBIrAKqxRGK0wGpRSPRORHgvOQeYgzYQ4ETInbNsU
sn2miNZQCjWDJa3mFv0dt39m1Rn1xC+unY1Trrp2Y8BgWdFoCeWiYqisKVjF2cWMxYuORksICoqgoC
gYMDoXG4B4cF5IHSSpkKTCQEmxatRwxYTFOaG+7GlG+fhySzh1NqUU6mN2atyudl5wIoBieCDPMgwU
5VBx3XRIIVB4A40YljNFqgpkuoS4XNTKJBgfEUhCyQpDIWgHWZI7ZCv2JJnCGPAC9WXPyICmGcmYXa
i6+rrVtlSpeaJY2Li2gHPQbAlaKYphewsaxciIZqQUwOgwlKdAjbdrsATL56FSzQ+m1ENbE1EiNNpr
lUNFnAqn5lLWTRU4X8nG7Z79y/V77hpa3YqE4QFNKVQUA4VWijgTzlccXiAsKoKSoTDoMNZhfIrSCa
AQn+GWHa7mSOuOpOVIIkEpKAWK0SENkpcJYGZDwHIknDidNOxSzdWaLc9ASbFxbQFrFKA4dTZlbjHj
jlsHwCgSr0hFk2GIKxl+YQl8PSdApehCihFDoWwpD2gCLSgvSAqPP1NjzYRlel2BUqjwPteKc6pqgX
qcwlLN8ez+JrdcV2Kp5mhFwlunUnZc7xkbs4ShJigGnJzNuLDYYPrmTYxPXUl+oFYROceLu49g45iZ
LUXCQQUCSgvO51vxrVMJCxXHweMxt99UIs58U4PUklTQGmbnMp7+V5PZsxkClIuKcknnMZTm+ImII6
8usXC6yt93HwECoAAUOHk84tiRKvVaxvP76jiB1AlzixljQ/l2STLh5FyKCFijSFOqVlCNJBMCq7rG
4X1uJUbDU3vqjI4YRkctb76bMFHO9/3p2QsszM6BApfVee/tC11jqlx07H6mhndCnOb3+imL9z1zMh
qixDesCNU4EYqB6jpad2LbzRoNT9RKqFZSxksFBEhaEQee29N1ynNLGar97AVakafQTqrjls5Ld11j
FEkqVauUaiSpYExvcucs6LfhfnvtvFxpz71n7/vG+yzbuXb2bQttxdLQSnINGE3X/73vCwLdQ0WpS8
B1pkluv6DaDORb8NLuyfl8La3z8SiiZqPEnY+7AHoa6D+IOpHE9wLS99ufJe0ElFKXvOita9oUNCNX
19WGnMt3QZ8IZeV/OxjyOfIBBqSvLtIu4eVK43xPgGkmJE7mdSTpS5mjYXRvYgepIqezB6zHDHRKJj
2m6DGo1cr3CHmPIHkizknTiPxVHz587vxVV5hvAc1O3bx0Gk2V095eo1dXYWXTkmfXS6AHpp+tDgOZ
kzOBUTe/cvjM+xrgvoeO7kp9ugNUpSPCruj8SvV3d0ofgpWs5Al0uiVWMAMCUb3pf/SFBw69CX1N6Z
e+d/RoQrbDOdnbWTtXvbSD9Ki+tFNaoZf+zqjDZBeLvCjCDb/584mdnfkrmtJfPfbuCeDzD96/5Zur
x8zPFaztpxp6AgPVC9zOOneiNivtxygRtGG+Uvc//c4jR3/LSrz/+9vwzjunw1unC1+dnU+/PlBUt4
2PmCDNhE3rCgyWNIdPJNy4Jeyy/N58SqmoWTWiOfBOTLXh01rd76vUZadtJk/+Zf/7l/0u+NCv4851
x9apATccfk6JumH7p8ItV662Vx58J151dsGHAM77WFAXbtkanvGeY6++uXyg5lr7X3ppsf5Ra/8XlG
cS6NzFDUIAAAAASUVORK5CYII=```

**Comment by Jyers** [comment]
It's certainly possible. I'll put it on my todo list to remind me to look into it :)

**Reply by alexqzd** [reply]
[I was able to display an image on the screen.](https://reddit.com/r/VoxelabAquila/comments/o9zyqz/just_a_little_sneak_peek_of_something_im_working/) But it requires a post processing script in PrusaSlicer to convert the thumbnail from PNG to a more DWIN friendly JPEG.

**Reply by alexqzd** [reply]
Here's the code: https://github.com/alexqzd/Marlin/commit/1d31b7770b8984b5f82962d7b602082e6169f214

**Reply by Jyers** [reply]
Nice work, would be interesting to see if we could somehow get a small version of the image into the file select / maybe also get cura gcode previews working.

Btw great work with the Aquila port Alex! If you ever want to chat about features of anything email me at Jyers.Marlin@gmail.com and we can find a good means of communication :)

**Comment by Giadej** [comment]
Another developer has already successfully implemented this feature on Ender3v2
![curae3v2](https://user-images.githubusercontent.com/80910217/125058438-8d9ef880-e0aa-11eb-89a5-6f5add6ae3b1.jpeg)

**Comment by Giadej** [comment]
@tititopher68-dev Developer mriscoc did it in his experimental build.
There is my video-preview in the OP of this link:
https://www.reddit.com/r/ender3v2/comments/ogb0ul/whats_new_on_the_experimental_version_of_the/

**Reply by mriscoc** [reply]
> @Giadej
> Thanks !
> I should have known ... So with him, the sources will eventually be available ... in 4 months. I had succeeded once in **intercepting** a PR on one of these "experimental" functions, I had asked a question and presto! Everything has been deleted ...
> **Whether he takes all or part of the code of one or the other**, OK, but that he cites the sources and that he does not make them private as he does. What is appreciable in the community here around the Firmware of @Jyers, is that everything is available on everyone's Github ... except for mriscoc (except for a very basic branch and without these features) ...
> 

I received an email alerting me that I was invoked here. The sources are in my repository, and I do PR to the official Bugfix Marlin repo. When the PR are approved, then the sources are available for anybody through an official source, that contribute to reduce the fragmentation. I applaud the work of JYers and all others real developers. I did not "put or delete" things because you, in fact, I don't remember you, I adapt my code by request of the great group of people that are part of the community of our Telegram group. Continue with you great work and be happy.

**Reply by Giadej** [reply]
> 
> 
> @Giadej
> You too attribute to yourself the fruit of Gotcha's labor and you give it to this mriscoc!
> Lamentable!

Relax yourself young boy, you're embarrassing with your hate shooting on everyone!!!
I gave everyone the right credits! 

**AngelV** for "Main Panel" icons
**Lighty1989** for "XYZ Move Axis" icons and "Bed Points" icons
**Gotcha's** for other icons
**Giadej** for custom Boot-Screen

https://github.com/mriscoc/Marlin_Ender3v2/blob/Ender3v2-Released/display%20assets/Giadej%20compilation/readme.md

**Reply by Giadej** [reply]
> I am only targeting Mriscoc (and I am not the only one) who does not respect the Marlin license law and who does not share anything and who took code from me by attributing it paternity. Ok ! It is not correct. Let him comply with the licensing rules and the publication of his real sources and their authors and all this will stop immediately. We have already asked him politely with courtesy the first few times and he answers us by banishment and by taking us for imbeciles (to be correct). Here you take advantage of free access to the code of Jyers and other contributors. Do you find it normal that Mriscoc flouts the licenses, takes advantage of our work and provides no corresponding sources for the binaries he publishes?
> Know one thing, Reddit moderation contacted me and are now taking this all seriously since I provided Reddit moderation with factual material.
> By doing this, Mriscoc kills the very principles of Open Source under the GPL license ... All this because he is looking for money a priori. You only need to see his ad to go to Telegram. Many people gave donations thinking that it was the way to obtain its firmware ... especially that Mriscoc does not give anything in the end in return ... This also raises questions for the moderation of Reddit ...
> Hoping that Google translate does not alter what I write ....

Sorry but from the way you posed with malice it seemed that your attack was referring to me. 
These developer wars aren't good for end users like me. We use your work and thank you for the work done. 
As regards the licensing problem, there are checks in progress and one has to wait to understand any discrepancies. 
As for donations, I can only tell you that you are wrong: all experimental releases on mriscoc's Telgram channel are FREE. 
The end user is free to support the developer with donations.

P.S.: Also for me "Hoping that Google translate does not alter what I write ...."
